From 10fb8f1a83b7fdb443cd821c234249b0e78c7a6e Mon Sep 17 00:00:00 2001
From: caiyinyu <caiyinyu@loongson.cn>
Date: Wed, 28 Jul 2021 10:29:48 +0800
Subject: [PATCH] glibc-2.28: delete __tp in struct __jmp_buf_internal_tag and
 modify related files.

        sysdeps/loongarch/bits:
                setjmp.h: delete __tp in struct __jmp_buf_internal_tag

        sysdeps/loongarch:
                __longjmp.S: modified to adjust new struct __jmp_buf_internal_tag
                setjmp.S: Likewise

        sysdeps/unix/sysv/linux/loongarch/lp64:
                jmp_buf-macros.h: Likewise

Change-Id: I17cc4b28d2e793f061c7420a60e2bd3d250fcf1d
---
 sysdeps/loongarch/__longjmp.S                 | 42 +++++++++----------
 sysdeps/loongarch/bits/setjmp.h               |  4 +-
 sysdeps/loongarch/setjmp.S                    | 42 +++++++++----------
 .../linux/loongarch/lp64/jmp_buf-macros.h     |  8 ++--
 4 files changed, 47 insertions(+), 49 deletions(-)

diff --git a/sysdeps/loongarch/__longjmp.S b/sysdeps/loongarch/__longjmp.S
index f556157371..68f6763954 100644
--- a/sysdeps/loongarch/__longjmp.S
+++ b/sysdeps/loongarch/__longjmp.S
@@ -21,27 +21,27 @@
 
 ENTRY (__longjmp)
   REG_L ra, a0, 0*SZREG
-  REG_L sp, a0, 2*SZREG
-  REG_L x, a0, 3*SZREG
-  REG_L fp, a0, 4*SZREG
-  REG_L s0, a0, 5*SZREG
-  REG_L s1, a0, 6*SZREG
-  REG_L s2, a0, 7*SZREG
-  REG_L s3, a0, 8*SZREG
-  REG_L s4, a0, 9*SZREG
-  REG_L s5, a0, 10*SZREG
-  REG_L s6, a0, 11*SZREG
-  REG_L s7, a0, 12*SZREG
-  REG_L s8, a0, 13*SZREG
-
-  FREG_L $f24, a0, 14*SZREG + 0*SZFREG
-  FREG_L $f25, a0, 14*SZREG + 1*SZFREG
-  FREG_L $f26, a0, 14*SZREG + 2*SZFREG
-  FREG_L $f27, a0, 14*SZREG + 3*SZFREG
-  FREG_L $f28, a0, 14*SZREG + 4*SZFREG
-  FREG_L $f29, a0, 14*SZREG + 5*SZFREG
-  FREG_L $f30, a0, 14*SZREG + 6*SZFREG
-  FREG_L $f31, a0, 14*SZREG + 7*SZFREG
+  REG_L sp, a0, 1*SZREG
+  REG_L x, a0,  2*SZREG
+  REG_L fp, a0, 3*SZREG
+  REG_L s0, a0, 4*SZREG
+  REG_L s1, a0, 5*SZREG
+  REG_L s2, a0, 6*SZREG
+  REG_L s3, a0, 7*SZREG
+  REG_L s4, a0, 8*SZREG
+  REG_L s5, a0, 9*SZREG
+  REG_L s6, a0, 10*SZREG
+  REG_L s7, a0, 11*SZREG
+  REG_L s8, a0, 12*SZREG
+
+  FREG_L $f24, a0, 13*SZREG + 0*SZFREG
+  FREG_L $f25, a0, 13*SZREG + 1*SZFREG
+  FREG_L $f26, a0, 13*SZREG + 2*SZFREG
+  FREG_L $f27, a0, 13*SZREG + 3*SZFREG
+  FREG_L $f28, a0, 13*SZREG + 4*SZFREG
+  FREG_L $f29, a0, 13*SZREG + 5*SZFREG
+  FREG_L $f30, a0, 13*SZREG + 6*SZFREG
+  FREG_L $f31, a0, 13*SZREG + 7*SZFREG
 
   sltui a0,a1,1
   add.d  a0, a0, a1   # a0 = (a1 == 0) ? 1 : a1
diff --git a/sysdeps/loongarch/bits/setjmp.h b/sysdeps/loongarch/bits/setjmp.h
index 92711def51..cc9b6bfda9 100644
--- a/sysdeps/loongarch/bits/setjmp.h
+++ b/sysdeps/loongarch/bits/setjmp.h
@@ -23,11 +23,9 @@ typedef struct __jmp_buf_internal_tag
   {
     /* Program counter.  */
     long int __pc;
-    /* TLS */
-    long int __tp;
     /* Stack pointer.  */
     long int __sp;
-    /* Global pointer*/
+    /* Reserved */
     long int __x;
     /* Frame pointer. */
     long int __fp;
diff --git a/sysdeps/loongarch/setjmp.S b/sysdeps/loongarch/setjmp.S
index d5a7dd4fa1..fbea3827a9 100644
--- a/sysdeps/loongarch/setjmp.S
+++ b/sysdeps/loongarch/setjmp.S
@@ -28,27 +28,27 @@ ENTRY (setjmp)
 END (setjmp)
 ENTRY (__sigsetjmp)
   REG_S ra, a0, 0*SZREG
-  REG_S sp, a0, 2*SZREG
-  REG_S x, a0, 3*SZREG
-  REG_S fp, a0, 4*SZREG
-  REG_S s0, a0, 5*SZREG
-  REG_S s1, a0, 6*SZREG
-  REG_S s2, a0, 7*SZREG
-  REG_S s3, a0, 8*SZREG
-  REG_S s4, a0, 9*SZREG
-  REG_S s5, a0, 10*SZREG
-  REG_S s6, a0, 11*SZREG
-  REG_S s7, a0, 12*SZREG
-  REG_S s8, a0, 13*SZREG
-
-  FREG_S $f24, a0, 14*SZREG + 0*SZFREG
-  FREG_S $f25, a0, 14*SZREG + 1*SZFREG
-  FREG_S $f26, a0, 14*SZREG + 2*SZFREG
-  FREG_S $f27, a0, 14*SZREG + 3*SZFREG
-  FREG_S $f28, a0, 14*SZREG + 4*SZFREG
-  FREG_S $f29, a0, 14*SZREG + 5*SZFREG
-  FREG_S $f30, a0, 14*SZREG + 6*SZFREG
-  FREG_S $f31, a0, 14*SZREG + 7*SZFREG
+  REG_S sp, a0, 1*SZREG
+  REG_S x, a0,  2*SZREG
+  REG_S fp, a0, 3*SZREG
+  REG_S s0, a0, 4*SZREG
+  REG_S s1, a0, 5*SZREG
+  REG_S s2, a0, 6*SZREG
+  REG_S s3, a0, 7*SZREG
+  REG_S s4, a0, 8*SZREG
+  REG_S s5, a0, 9*SZREG
+  REG_S s6, a0, 10*SZREG
+  REG_S s7, a0, 11*SZREG
+  REG_S s8, a0, 12*SZREG
+
+  FREG_S $f24, a0, 13*SZREG + 0*SZFREG
+  FREG_S $f25, a0, 13*SZREG + 1*SZFREG
+  FREG_S $f26, a0, 13*SZREG + 2*SZFREG
+  FREG_S $f27, a0, 13*SZREG + 3*SZFREG
+  FREG_S $f28, a0, 13*SZREG + 4*SZFREG
+  FREG_S $f29, a0, 13*SZREG + 5*SZFREG
+  FREG_S $f30, a0, 13*SZREG + 6*SZFREG
+  FREG_S $f31, a0, 13*SZREG + 7*SZFREG
 
 #if !IS_IN (libc) && IS_IN(rtld)
   li v0, 0
diff --git a/sysdeps/unix/sysv/linux/loongarch/lp64/jmp_buf-macros.h b/sysdeps/unix/sysv/linux/loongarch/lp64/jmp_buf-macros.h
index 5cefbf4be0..e1c96e67c4 100644
--- a/sysdeps/unix/sysv/linux/loongarch/lp64/jmp_buf-macros.h
+++ b/sysdeps/unix/sysv/linux/loongarch/lp64/jmp_buf-macros.h
@@ -33,9 +33,9 @@
        printf ("#define SAVED_MASK_OFFSET %d\n", offsetof (struct __jmp_buf_tag, __saved_mask));
    } */
 
-# define JMP_BUF_SIZE 312
+# define JMP_BUF_SIZE 304
 # define JMP_BUF_ALIGN 8
-# define SIGJMP_BUF_SIZE 312
+# define SIGJMP_BUF_SIZE 304
 # define SIGJMP_BUF_ALIGN 8
-# define MASK_WAS_SAVED_OFFSET 176
-# define SAVED_MASK_OFFSET 184
+# define MASK_WAS_SAVED_OFFSET 168
+# define SAVED_MASK_OFFSET 176
-- 
2.27.0

